### PR TITLE
Add support for partial path merging with path exploration

### PR DESCRIPTION
**Pull request by karkhaz**
This PR depends on #1641; please review only the last commit in the PR (`Able to do partial merging with path exploration`).

This commit adds a new dummy function `__CPROVER_begin_path_explore()` and new flag `--partial-merge`. When both `--paths` and `--partial-merge` are passed to CBMC, goto-symex will perform path-merging until it encounters a call to `__CPROVER_begin_path_explore()`. From that point onward, it will do path exploration (saving untaken paths onto a workqueue and not merging paths).

This mode is useful where we don't want to resume execution from boring code early on in `main()`. The idea is that we should do path-merging along all initialization code, and then do path-exploration from that point onward.

### CODE REVIEW

File: src/cbmc/bmc.h [hunk 1]
```diff
@@ -294,6 +296,8 @@ class path_explorert:public bmct
   " --unwind nr                  unwind nr times\n" \
   " --unwindset L:B,...          unwind loop L with a bound of B\n" \
   "                              (use --show-loops to get the loop IDs)\n"  \
+  " --partial-merge              do path exploration only after executing\n" \
+  "                              __CPROVER_begin_path_explore()\n"  \
```
Comment (smowton) [inline src/cbmc/bmc.h]: Perhaps "encountering __CPROVER_begin_path_explore in the input program" would be clearer

File: src/cbmc/bmc.h [hunk 2]
```diff
@@ -294,6 +296,8 @@ class path_explorert:public bmct
   " --unwind nr                  unwind nr times\n" \
   " --unwindset L:B,...          unwind loop L with a bound of B\n" \
   "                              (use --show-loops to get the loop IDs)\n"  \
+  " --partial-merge              do path exploration only after executing\n" \
```
Comment (smowton) [inline src/cbmc/bmc.h]: This feels like another `--paths` option, merge the two so `--paths partial` works?

**Review by smowton — COMMENTED** [review]
Also should we provide a corresponding start-merging-again flag?

File: src/goto-programs/safety_checker.h [hunk 3]
```diff
@@ -45,4 +48,41 @@ class safety_checkert:public messaget
...
+{
+  switch(a)
+  {
+  case safety_checkert::resultt::ERROR:
+    return a;
+  case safety_checkert::resultt::SAFE:
+    a=b;
+    return a;
+  case safety_checkert::resultt::UNSAFE:
+    a = b == safety_checkert::resultt::ERROR ? b : a;
```
Comment (thk123) [inline src/goto-programs/safety_checker.h]: Style: this line is sort of hard to parse - consider breaking it out:
```
bool b_is_error = b == safety_checkert::resultt::ERROR;
a = b_is_error ? b : a;

File: src/goto-programs/safety_checker.h [hunk 4]
```diff
@@ -45,4 +48,41 @@ class safety_checkert:public messaget
   const namespacet &ns;
 };
 
+/// \brief The worst of two results
```
Comment (thk123) [inline src/goto-programs/safety_checker.h]: Consider elaborating these comments to explain what best and worst mean.

File: src/analyses/dirty.h [hunk 5]
```diff
@@ -17,42 +17,74 @@ Date: March 2013
...
   }
 
-  explicit dirtyt(const goto_functionst &goto_functions)
+  explicit dirtyt(const goto_functionst &goto_functions):
+    initialized(false)
   {
-    forall_goto_functions(it, goto_functions)
-      build(it->second);
+    build(goto_functions);
+    // build(g_funs) responsible for setting initialized to true, since
```
Comment (thk123) [inline src/analyses/dirty.h]: This possibly isn't your design - but it would seem to be better for this constructor only to be visible, then you don't need to separately track initialised and have constructors as the only way build this object (of course removing the default constructor).

**Review by thk123 — COMMENTED** [review]
Just reviewed the `src/analyses` and `scripts/` stuff. I think it would be much better use constructors to enforce correct construction of `dirtyt` so you don't need run time checks.

**Comment by thk123** [comment]
Urgh sorry for putting the review in the wrong place, should have read more carefully, will migrate comments over to the real PR.

**Comment by martin-cs** [comment]
@karkhaz Is this one next on the list?

**Comment by karkhaz** [comment]
@martin-cs yes. I'm a bit busy for the next two days, but I'll rebase and address all comments soon

**Comment by martin-cs** [comment]
@karkhaz Sure; shout when you want it reviewed.

**Comment by karkhaz** [comment]
@martin-cs @smowton rebased and implemented your suggestions, thank you! wrt your suggestion of a start-merging-again flag: @tautschnig and I discussed this briefly. I don't mind implementing it, but I think that we came to the conclusion that this functionality could easily become confusing, e.g. if you put these `__CPROVER_{begin,end}_path_explore` calls in function calls that are called repeatedly, or in loops, or if you have more than one pair in a program. I suppose you could argue that it's the user's responsibility to deal with the confusion; I don't really mind either way.

The concrete use case we had was to put a single `__CPROVER_begin_path_explore` at some point in some straight-line code in `main()`, the idea being to path-merge through some boring initialization code in `main()` and then do path exploration from that point onwards---which seems like a fairly straightforward way to use this feature.

**Review by martin-cs — APPROVED** [review]
I'm happy that the code implements this functionality correctly.  I think I'd like to have a quick discussion over whether this is the best way of approaching this.

**Comment by martin-cs** [comment]
@karkhaz It does seem a good use of this feature; I'm just concerned about whether focusing on this case and implementing it means we miss a slightly different implementation that has many other use cases.  Maybe @DanielNeville @cesaro @SeanHeelan @johnfxgalea @polgreen @theyoucheng could comment on their uses of path merging?

**Comment by karkhaz** [comment]
@martin-cs I'm interested to hear other use cases. FWIW, my eventual vision is that this becomes automatic. This could be static or dynamic. Static means that we do some sort of scan over the goto-program to figure out the 'shape' of the code, and use heuristics to figure out where in the code would be most productive to merge, and where would be most productive to explore. @tautschnig suggested that it could be worthwhile to do this _during_ symbolic execution also: if we're path-exploring a loop and we spin round several times and nothing much seems to have changed, then we could decide to merge and move on to more interesting things.

I posted this for review anyway because:
- I will need to use it manually through some test codebases, in order to develop heuristics that work automatically and see how effective they are;
- I suppose the heuristics are necessarily going to be very conservative, and it's quite possible that somebody with domain knowledge of their codebase will want to place this macro manually in a place that the heuristic misses.

**Comment by karkhaz** [comment]
Oh, and another reason: I don't actually have an ETA for implementing automatic decisions on when to merge. I've got about 6 months of PhD time left, and implementing that is one of many possible things I could be doing, many of which won't get done :slightly_smiling_face: just in case I was giving the impression that this was something I was interested in doing imminently.

**Comment by NlightNFotis** [comment]
Hello.

This PR appears to be in a stale state (a lot of merge conflicts and outdated CI information). I am closing this PR now, but should you feel like you want this PR merged, feel free to reopen the PR and rebase on branch `develop` to continue this work.
